fix: add more tests for curriclum testing

moT01 opened the pull request:
PR for https://github.com/freeCodeCamp/freeCodeCamp/pull/38456#issuecomment-604825040

Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] All the files I changed are in the same world language, for example: only English changes, or only Chinese changes, etc.


This PR adds two tests to check that a challenge title and id are found in their meta.json files.
In making these tests, I found one challenge where a title didn't match.
It also adds some options to test a specific block or superblock of the curriculum...
`npm run test:curriculum --block='Basic HTML and HTML5'`
`npm run test:curriculum --superblock=responsive-web-design`
These commands will do that.

The lines I actually changed in `test-challenges.js` are 35, 103-131, & 205-238. The rest of the diff wasn't changed other than some indentation. 

The code feels a little sloppy, but it looks like it's working. Any suggestions are welcome.

moT01 committed:
fix: es-lint error

moT01 committed:
fix: formatting in doc

moT01 committed:
fix: use implicit return to save code

RandellDawson commented on curriculum/getChallenges.js:
```diff
@@ -16,10 +16,9 @@ function getChallengesDirForLang(lang) {
 function getMetaForBlock(block) {
-  const meta = JSON.parse(
+  return JSON.parse(
     fs.readFileSync(path.resolve(metaDir, `./${block}/meta.json`), 'utf8')
```
Shouldn't we be use readFile instead of readFileSync?

moT01 replied:
not sure, should we?

RandellDawson replied:
It would be non-blocking using the async readFile.

moT01 replied:
Don't we want to wait for that to finish so we're sure the data is there. I dunno, I tried to switch to the async method and it keeps giving me an error - presumably because the data hasn't been retrieved before it tries to access it.

ojeytonwilliams replied:
The problem seems to be that `getMetaForBlock` is called inside `describe` and `describe` only takes synchronous functions as its arguments.  I think, anyway.

The solution seems to be to do all the async setup first, then run the tests.  Currently it's a mixture.  If we do that then we can potentially save a bit of time by starting puppeteer and sourcing the challenges in parallel.  At the very least it should be easier to reason about.

moT01 replied:
So are you guys thinking this needs to be changed up?

ojeytonwilliams committed:
refactor: move all async code to setup

Also simplifies cleanup and renames challenges. It used to be an object
containing a challenges property which was a little confusing.

ojeytonwilliams commented on curriculum/test/test-challenges.js:
```diff
@@ -86,59 +87,115 @@ spinner.text = 'Populate tests.';
+  // the next few statements will filter challenges based on command variables
+  if (process.env.npm_config_superblock) {
+    langAndChallenges[0].challenges = langAndChallenges[0].challenges.filter(
+      challenge => challenge.superBlock === process.env.npm_config_superblock
```
Could this be looser?  A case insensitive regex would be easy enough to write and is a touch more user-friendly.

moT01 replied:
Working on it.

ojeytonwilliams commented on curriculum/test/test-challenges.js:
```diff
@@ -86,59 +87,115 @@ spinner.text = 'Populate tests.';
+
+  if (process.env.npm_config_block) {
+    langAndChallenges[0].challenges = langAndChallenges[0].challenges.filter(
+      challenge => challenge.block === process.env.npm_config_block
```
Same question as for superBlocks.

ojeytonwilliams committed:
refactor: rename langAndChallenges

ojeytonwilliams commented:
Fixed my awful naming with @moT01's much better one.

moT01 commented on curriculum/test/test-challenges.js:
```diff
@@ -86,59 +87,115 @@ spinner.text = 'Populate tests.';
-  );
+  // the next few statements will filter challenges based on command variables
+  if (process.env.npm_config_superblock) {
+    langAndChallenges[0].challenges = langAndChallenges[0].challenges.filter(
```
We probably don't need to test any languages that aren't live anywhere. I'm really not sure what's all out there, I think Chinese is for sure - not sure on the other ones.

ojeytonwilliams replied:
Perhaps not, but the tests should fail with an explanation if we aren't supporting this.

moT01 committed:
fix: add flexibility for input strings

RandellDawson commented on curriculum/test/test-challenges.js:
```diff
@@ -125,6 +130,57 @@ async function setup() {
+      challenge => challenge.superBlock === filter
+    );
+
+    if (challengesForLang[0].challenges.length === 0) {
```
```suggestion
    if (!challengesForLang[0].challenges.length) {
```

moT01 committed:
fix: comparison syntax

moT01 committed:
fix: update docs

ojeytonwilliams commented on curriculum/test/test-challenges.js:
```diff
@@ -86,59 +88,141 @@ spinner.text = 'Populate tests.';
+  global.Worker = createPseudoWorker(await newPageContext(browser));
+  page = await newPageContext(browser);
+  await page.setViewport({ width: 300, height: 150 });
+  const testLangs = testedLangs();
```
```suggestion
  const testLangs = testedLangs();
  if (testLangs.length > 1)
    throw Error(
      `Testing more than one language at once is not currently supported
please change the TEST_CHALLENGES_FOR_LANGS env variable to a single language`
    );
```
We may not need to support multiple languages, but we should tell people why it fails.

ojeytonwilliams commented on curriculum/test/test-challenges.js:
```diff
@@ -86,59 +88,141 @@ spinner.text = 'Populate tests.';
-        await browser.close();
+  // the next few statements create a list of all blocks and superblocks
+  // as they appear in the list of challenges
+  const filters = {
+    blocks: [],
+    superBlocks: []
+  };
+
+  challengesForLang[0].challenges.forEach(chal => {
+    if (filters.blocks.indexOf(chal.block) < 0) {
+      filters.blocks.push(chal.block);
+    }
+
+    if (filters.superBlocks.indexOf(chal.superBlock) < 0) {
+      filters.superBlocks.push(chal.superBlock);
+    }
+  });
```
```suggestion
  const blocks = challengesForLang[0].challenges.map(({ block }) => block);
  const superBlocks = challengesForLang[0].challenges.map(
    ({ superBlock }) => superBlock
  );
  const targetBlockStrings = [...new Set(blocks)];
  const targetSuperBlockStrings = [...new Set(superBlocks)];
```
This is a little more concise and is a more standard way (I think!) of removing duplicates.

ojeytonwilliams commented on curriculum/test/test-challenges.js:
```diff
@@ -86,59 +88,141 @@ spinner.text = 'Populate tests.';
+  if (process.env.npm_config_superblock) {
+    const filter = stringSimilarity.findBestMatch(
+      process.env.npm_config_superblock,
+      filters.superBlocks
```
```suggestion
      targetSuperBlockStrings
```
`filter` was a little overloaded.

ojeytonwilliams commented on curriculum/test/test-challenges.js:
```diff
@@ -86,59 +88,141 @@ spinner.text = 'Populate tests.';
+  if (process.env.npm_config_block) {
+    const filter = stringSimilarity.findBestMatch(
+      process.env.npm_config_block,
+      filters.blocks
```
```suggestion
      targetBlockStrings
```
Same again.

ojeytonwilliams reviewed (changes requested):
@moT01 I found a few minor things.  I've tested it locally and it looks good, though.

moT01 committed:
Update curriculum/test/test-challenges.js

Co-Authored-By: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

moT01 committed:
Update curriculum/test/test-challenges.js

Co-Authored-By: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

moT01 committed:
Update curriculum/test/test-challenges.js

Co-Authored-By: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

moT01 committed:
Update curriculum/test/test-challenges.js

Co-Authored-By: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

ojeytonwilliams reviewed (approved):
LGTM
